### ISSUE TITLE
Feature: Ability to Drop columns

### ISSUE THREAD

**Issue by bputt** [crate/crate#4457]
**Use case**: Removing columns that aren't needed any more or to allow removal of columns that were accidentally added.

New columns could be added very easily, typos for example:

```
UPDATE mytable SET someMispelledColumn = ?
```

**Feature description**: Allow us to drop columns using the ALTER command, so for example:

```
ALTER TABLE mytable DROP COLUMN mycolumn;
```

## Possible Constraints

- Data is not deleted from disk
- Cannot re-use same column name with different type (column is marked as deleted instead of removed)
  - Would have to introduce oids for columns and de-couple column name from storage to allow this

**Comment by celaus** [comment]
Hi @bputt - thanks! We already have that in our backlog, I'll add your upvote to it 👍

**Comment by cweiske** [comment]
Schema migrations to correct errors that we made yesterday(tm) are currently not possible at all.
A "migration" currently means: Drop the table and re-add it :/

**Comment by anottrott** [comment]
Would definitely be useful!

**Comment by diyism** [comment]
From a developer's perspective, both array _ count(https://github.com/crate/crate/issues/882)  and drop column(https://github.com/crate/crate/issues/4457) are the most important and relatively easy to implement features

**Comment by mfussenegger** [comment]
> From a developer's perspective, both array _ count(#882) and drop column(#4457) are the most important and relatively easy to implement features

If you think those are easy to implement, we're happy to accept contributions and can also give you some pointers how to do that if you're interested.

A slow version of `array_count` would indeed be rather simple. All one had to do is to implement a new scalar function. Taking a look at one of the PRs that added other scalar functions should be a good starting point.
The reason we didn't add it already, other than lack of time, is that a common use case for it is to query for all records where "an array is not empty" or "an array has more than 1 element".
Implementing `array_count` as a simple scalar function would lead in a slow query execution and we'd need to introduce an option for an optimized query generation for this common use-case. Otherwise we wouldn't meet the expectations most people likely have. This makes this "not easy".

Drop column is even worse once you ask yourself questions like "what to do with the already indexed data" and "what to do if a column with the same name but different type is added later on". Possible answers to these questions are not easy.

**Comment by mkleen** [comment]
This patch https://issues.apache.org/jira/browse/LUCENE-1812 provides the ability to remove stored fields. It was never ported over to Lucene 4.x but it might be worth looking into the details. https://github.com/apache/lucene/issues/4990

**Comment by matriv** [comment]
Remaining tasks for https://github.com/crate/crate/tree/drop-column
```
 - [x] try to simulate concurrency (e.g drop a column while it has been used in a new column with a generated expression in between)
 - [ ] dropping parent drops subcols
 - [ ] information_schema,pg_catalog 
 - [ ] check all places that the dropped columns are not "seen" (views, show create table, select _raw, etc)
 - [ ] use oid (already assigned in mapping) instead of colname
 - [ ] re-add column with same name

**Comment by BaurzhanSakhariev** [comment]
Implemented in https://github.com/crate/crate/pull/14824 and will be available in 5.5.0 release.


>Possible Constraints

>Data is not deleted from disk
>Cannot re-use same column name

Data deletion is tracked by https://github.com/crate/crate/issues/14254.

It's possible to re-use same column name (also with different type).
